### PR TITLE
Scaled Exponential Linear Unit (SELU) activation

### PR DESCRIPTION
**Pull request by vipulraheja**
[Self-Normalizing Neural Networks](https://arxiv.org/abs/1706.02515)
 - [x] A new activation function called SELU
 - [x] Unit test

### CODE REVIEW

**Comment by vipulraheja** [comment]
The build failures seem to be unrelated to the change, since the respective test [passed](https://travis-ci.org/Theano/Theano/jobs/244119050) in TravisCI. Seems most likely related to https://github.com/Theano/Theano/issues/6048

**Comment by vipulraheja** [comment]
Looks like the changes in #6062 fixed it.

**Comment by vipulraheja** [comment]
@nouiz @lamblin Could you please take a look whenever you get a chance? Would love some feedback :)

File: theano/tensor/nnet/nnet.py [hunk 1]
```diff
@@ -2448,6 +2448,29 @@ def elu(x, alpha=1):
...
+    -------
+    symbolic tensor
+        Element-wise scaled exponential linear activation function applied to `x`.
+
+    References
+    ----------
+    .. [1] Klambauer G, Unterthiner T, Mayr A, Hochreiter S.
+        "Self-Normalizing Neural Networks" <https://arxiv.org/abs/1706.02515>
+    """
+    alpha = 1.6732632423543772848170429916717
```
Comment (nouiz) [inline theano/tensor/nnet/nnet.py]: Here you override the alpha from the function signature. If you want this value to be the default, set it in the fct signature.

A quick look at the other framework interface make me think we should just remove the alpha parameter from the fct signature.

Comment (vipulraheja) [inline theano/tensor/nnet/nnet.py]: You are absolutely right. It was indeed a residual argument. Removed it.

File: theano/tensor/nnet/nnet.py [hunk 2]
```diff
@@ -2448,6 +2448,29 @@ def elu(x, alpha=1):
     return tensor.switch(x > 0, x, alpha * tensor.expm1(x))
 
 
+def selu(x):
```
Comment (nouiz) [inline theano/tensor/nnet/nnet.py]: Add that in the doc in the file doc/library/tensor/nnet/nnet.txt. Add elu at the same time.

After that, I think the PR will be ready to merge.

thanks

Comment (vipulraheja) [inline theano/tensor/nnet/nnet.py]: Done and done.

File: theano/tensor/nnet/nnet.py [hunk 3]
```diff
@@ -2448,6 +2448,31 @@ def elu(x, alpha=1):
...
+        Tensor to compute the activation function for.
+
+    Returns
+    -------
+    symbolic tensor
+        Element-wise scaled exponential linear activation function applied to `x`.
+
+    References
+    ----------
+    .. [1] Klambauer G, Unterthiner T, Mayr A, Hochreiter S.
```
Comment (nouiz) [inline theano/tensor/nnet/nnet.py]: This line see to raise an error in the doc gen:

/home/travis/build/Theano/Theano/theano/tensor/nnet/nnet.py:docstring of theano.tensor.nnet.selu:13: WARNING: Duplicate explicit target name: "1".

Maybe you can just remove the [1]?

Comment (vipulraheja) [inline theano/tensor/nnet/nnet.py]: Seems like citation numbers are [global](http://www.sphinx-doc.org/en/stable/rest.html#citations) in Sphinx. Fixed them all here.

**Comment by nouiz** [comment]
travis failed in python 3.3. I updated sphinx to the same version as python 2.7. Hopefully this will fix travis.

**Comment by vipulraheja** [comment]
@nouiz I was able to generate the documentation properly on a new environment, (both with Sphinx v1.3.1 and v1.5.1). Hopefully this should pass. :crossed_fingers: 
screen shot 2017-07-01 at 11 24 22 pm

**Comment by vipulraheja** [comment]
@nouiz Looks like a lot of libraries (particularly sphinx) don't like Python3.3. I tried it with Python2.7 and it works just fine (the issues with references also were fixed, with both versions of Sphinx - 1.3.1 and 1.5.1).

**Comment by nouiz** [comment]
We are having others problems with python 3.3. We are seeing if bumping the min requirement to 3.4 would fix them. If so, this would probably also fix the problems in this PR. So I would just wait for the result of that and probably just rebase when this is done.

**Comment by vipulraheja** [comment]
Sounds good. Thank you!

**Comment by nouiz** [comment]
Can you rebase this PR on top of the master of Theano? This will remove the merge commit (so history is cleaner) and it should fix problem you have.

thanks

**Comment by vipulraheja** [comment]
@nouiz Ran into some issues with sourcetree due to merge conflicts and ended up with duplicate commits. Could you please suggest a solution? One thing I can think of is a squashed merge.

**Comment by abergeron** [comment]
Can you do a proper rebase to avoid the merge commits?  Also after a rebase, you need to do `git push -f` Even if git advises you to do a merge, don't listen to it.

**Comment by abergeron** [comment]
I did a proper rebase.  Please check that all the changes you want are there.

**Comment by vipulraheja** [comment]
@abergeron Sorry was sick and didn't get a chance to follow up. Thanks for taking this up. Looks good! 👍

**Comment by vipulraheja** [comment]
@abergeron Thanks!

**Pull request merged**